src/LibraryManager.Vsix/Shared/LibraryCommandService.cs

```diff
@@ -0,0 +1,396 @@
+            Telemetry.TrackUserTask("restore", TelemetryResult.None, telResult.Select(i => new KeyValuePair<string, object>(i.Key, i.Value)).ToArray());
+        }
+
+        private void AddErrorsToErrorList(string projectName, string configFile, IEnumerable<ILibraryOperationResult> results)
+        {
+            var errorList = new ErrorList(projectName, configFile);
+            errorList.HandleErrors(results);
+        }
+
+        private async Task AddFilesToProjectAsync(string configFilePath, Project project, IEnumerable<ILibraryOperationResult> results, CancellationToken cancellationToken)
+        {
+            string cwd = Path.GetDirectoryName(configFilePath);
+            var files = new List<string>();
+
+            foreach (ILibraryOperationResult state in results)
+            {
+                if (state.Success)
+                {
+                    IEnumerable<string> absoluteFiles = state.InstallationState.Files
+                        .Select(file => Path.Combine(cwd, state.InstallationState.DestinationPath, file)
+                        .Replace('/', Path.DirectorySeparatorChar));
+                    files.AddRange(absoluteFiles.Where(file => !files.Contains(file)));
+                }
+            }
+
+            if (project != null)
+            {
+                var logAction = new Action<string, LogLevel>((message, level) => { Logger.LogEvent(message, level); });
+                await VsHelpers.AddFilesToProjectAsync(project, files, logAction, cancellationToken);
+            }
+        }
+
+        private CancellationToken RegisterCancellationToken(CancellationToken userCancellation)
+        {
+            CancellationTokenSource internalSource = new CancellationTokenSource();
+            _internalCancellationTokenSource = internalSource;
+
+            CancellationTokenSource linkedSource = CancellationTokenSource.CreateLinkedTokenSource(userCancellation);
+            _linkedCancellationTokenSource = linkedSource;
+
+            linkedSource.Token.Register(() =>
+            {
+                internalSource.Cancel();
+            });
+
+            return internalSource.Token;
+        }
+
+        public void CancelOperation()
+        {
+            if (_internalCancellationTokenSource != null)
+            {
+                if (!_internalCancellationTokenSource.IsCancellationRequested)
+                {
+                    _internalCancellationTokenSource.Cancel();
+                }
+                _internalCancellationTokenSource.Dispose();
+            }
+
+            if (_linkedCancellationTokenSource != null)
+            {
+                 _linkedCancellationTokenSource.Dispose();
+            }
+
+            _currentOperationTask = null;
+        }
+
+        public void Dispose()
+        {
+            _solution.UnadviseSolutionEvents(_solutionCookie);
+        }
+
+        public int OnAfterOpenProject(IVsHierarchy pHierarchy, int fAdded)
```

Comment:
>public int OnAfterOpenProject(IVsHierarchy pHierarchy, int fAdded) [](start = 8, length = 66)

Not sure if you guys use solution events anywhere else, but the WTE implementations of DefaultSolutionEvents are pretty nice and remove the need to define all the empty methods down here.

Reply:
You mean do the same of what they do there or take dependencies on WTE?